wp_api/src/request.rs

```diff
@@ -439,6 +449,175 @@ impl WpNetworkResponse {
+    }
+
+    pub fn get_http_auth_method(
+        &self,
+    ) -> Result<Option<HttpAuthMethod>, HttpAuthMethodParsingError> {
+        let header_value = self
+            .header_map
+            .inner
+            .get(http::header::WWW_AUTHENTICATE)
+            .and_then(|h_v| h_v.to_str().ok());
+
+        if let Some(header_value) = header_value {
+            let result = HttpAuthMethod::from_str(header_value);
+
+            return match result {
+                Ok(method) => Ok(Some(method)),
+                Err(e) => Err(e),
+            };
+        }
+
+        Ok(None)
+    }
+}
+
+struct HttpRetryDuration {
+    seconds: u64,
+}
+
+/// Parser based on https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After
+///
+impl FromStr for HttpRetryDuration {
+    type Err = WpApiError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Handle the simple case where the header is a number of seconds
+        if let Ok(seconds) = s.parse::<u64>() {
+            return Ok(HttpRetryDuration { seconds });
+        }
+
+        // Handle the case where the header is a date string
+        let parsed =
+            DateTime::parse_from_rfc2822(s).map_err(|_| WpApiError::RequestExecutionFailed {
+                status_code: None,
+                redirects: None,
+                reason: RequestExecutionErrorReason::MisconfiguredRateLimitError {},
+            })?;
+
+        let now = Utc::now();
+        let duration = parsed.signed_duration_since(now);
+        let seconds = duration.num_seconds();
+
+        Ok(HttpRetryDuration {
+            seconds: seconds as u64,
+        })
+    }
+}
+
+#[derive(PartialEq, Eq, Debug, Clone, uniffi::Record)]
+pub struct BasicAuthenticationDetails {
+    pub realm: Option<String>,
+    pub requires_utf8: bool,
+}
+
+#[derive(PartialEq, Eq, Debug, Clone, uniffi::Record)]
+pub struct DigestAuthenticationDetails {
+    pub realm: Option<String>,
+    pub nonce: String,
+    pub qop: String,
+    pub algorithm: Option<String>,
+    pub opaque: String,
+    pub requires_utf8: bool,
+}
+
+#[derive(PartialEq, Eq, Debug, Clone, uniffi::Enum)]
+pub enum HttpAuthMethod {
+    Basic(BasicAuthenticationDetails),
+    Digest(DigestAuthenticationDetails),
+    Other(String, HashMap<String, String>),
+}
+
+#[derive(PartialEq, Eq, Debug, Clone, uniffi::Error)]
+pub enum HttpAuthMethodParsingError {
+    MissingNonce,
+    MissingQop,
+    MissingAlgorithm,
+    MissingOpaque,
+    Unknown, // Some case we're not handling yet
+}
+
+/// Parser based on https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate
+///
+impl FromStr for HttpAuthMethod {
+    type Err = HttpAuthMethodParsingError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts = s.splitn(2, ' ').collect::<Vec<&str>>();
+
+        let method = parts[0];
+
+        let mut details = HashMap::new();
+        parts[1].split(',').for_each(|part| {
```

Comment:
It's possible that `parts` contain less than two elements. In that case, would `parts[0]` and `parts[1]` would panic? Maybe we should do some array length check beforehand?